R/tag_env.R

```diff
@@ -0,0 +1,1419 @@
+        addClass = function(class) {
+          rebuild()
+          tagQueryClassAdd(manuallySelected(), class)
+          self
+        },
+        #' * `$removeClass(class)`: Removes a set of class values from all of the selected elements.
+        removeClass = function(class) {
+          rebuild()
+          tagQueryClassRemove(manuallySelected(), class)
+          self
+        },
+        #' * `$hasClass(class)`: Determine whether the selected elements have a given class. Returns a vector of logical values.
+        hasClass = function(class) {
+          rebuild()
+          tagQueryClassHas(manuallySelected(), class)
+        },
+        #' * `$toggleClass(class)`: If the a class value is missing, add it. If a  class value already exists, remove it.
+        toggleClass = function(class) {
+          rebuild()
+          tagQueryClassToggle(manuallySelected(), class)
+          self
+        },
+
+        #' * `$addAttrs(...)`: Add named attributes to all selected children. Similar to [`tagAppendAttributes()`].
+        addAttrs = function(...) {
+          rebuild()
+          tagQueryAttrsAdd(manuallySelected(), ...)
+          # no need to rebuild(); already flattened in add attr function
+          self
+        },
+        #' * `$removeAttrs(attrs)`: Removes the provided attributes in each of the selected elements.
+        removeAttrs = function(attrs) {
+          rebuild()
+          tagQueryAttrsRemove(manuallySelected(), attrs)
+          self
+        },
+        #' * `$emptyAttrs()`: Removes all attributes in each of the selected elements.
+        emptyAttrs = function() {
+          rebuild()
+          tagQueryAttrsEmpty(manuallySelected())
+          self
+        },
+        #' * `$hasAttr(attr)`: Returns a vector whose values are whether the selected element contains the non-`NULL` attribute.
+        hasAttr = function(attr) {
+          rebuild()
+          tagQueryAttrHas(manuallySelected(), attr)
+        },
+
+        #' ## Adjust child elements
+        #' * `$append(...)`: Add all `...` objects as children **after** any existing children to the selected elements. Similar to [`tagAppendChildren()`]
+        append = function(...) {
+          rebuild()
+          tagQueryChildrenAppend(manuallySelected(), ...)
+          rebuild()
+          self
+        },
+        #' * `$prepend(...)`: Add all `...` objects as children **before** any existing children to the selected elements. A variation of [`tagAppendChildren()`]
+        prepend = function(...) {
+          rebuild()
+          tagQueryChildrenPrepend(manuallySelected(), ...)
+          rebuild()
+          self
+        },
+        #' * `$empty()`: Remove all children in the selected elements. Use this method before calling `$append(...)` to replace all selected elements' children.
+        empty = function() {
+          rebuild()
+          tagQueryChildrenEmpty(manuallySelected())
+          # no need to rebuild
+          self
+        },
```

Comment:
Maybe these sort of functions (that operate on a selection) should warn when there is no selection (and inside devmode?)

Reply:
I think there are some times where one would legitimately want to allow empty selections. For example, suppose you want to write a function that finds tags with class `foo` and adds the class `bar`. You'd do something like:

```R
ensure_foobar <- function(x) {
  tagQuery(x)$find('.foo')$addClass('bar')$asTags()
}
```

But you wouldn't want this to throw a warning if the `.foo` selection was empty.

It might make sense to add an argument that tells it to warn on empty selections, but I think that's something we should come back to later.

Reply:
I think a dev mode warning is fair if you have not called `$find(".foo")` but are trying to call `$addClass("bar")` right away. 

I'm perfectly fine to not warn on a (user found) empty selection.

Reply:
Such as "Please call `$find()` before calling `$addClass()`

Reply:
I mean that it would be a general purpose function whose purpose is to ensure that the `bar` class is there along with `foo`, but the input may or may not have `foo`. Maybe this will illustrate better: imagine that you write a utility function that finds all `b` tags and replaces them with `i`, or one that capitalizes all text in `a()` tags, or something like that. Functions like that those should not throw warnings. Here's an analogy: imagine there was a global option to make `lapply` emit a warning if the input is length 0. That would be weird and emit unnecessary (and confusing) warnings.

Reply:
We now select all of the top level items by default. This avoids confusion all around.